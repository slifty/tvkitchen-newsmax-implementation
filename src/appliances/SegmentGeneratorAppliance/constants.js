export const INTERVALS = {
	INFINITE: 0,
	MINUTE: 60000,
	HOUR: 3600000,
	DAY: 86400000,
	WEEK: 604800000,
}
