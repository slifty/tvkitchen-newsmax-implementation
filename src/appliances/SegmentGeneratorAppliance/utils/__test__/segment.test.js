import {
	calculateStartingPosition,
	generateSegmentPayloadsFromPosition,
	generateSegmentPayloadsForPositionRange,
	getPeriodPosition,
} from '../segment'

import { INTERVALS } from '../../constants'

describe('segment utils #unit', () => {
	describe('calculateStartingPosition', () => {
		it('should return zero when passed matching timestamps', () => {
			expect(calculateStartingPosition(
				'2021-03-12T20:04:25.258Z',
				'2021-03-12T20:04:25.258Z',
			)).toBe(0)
		})
		it('should generate negative positions if startAt is before the reference origin', () => {
			expect(calculateStartingPosition(
				'2021-03-12T20:04:24.258Z',
				'2021-03-12T20:04:25.258Z',
			)).toBe(-1000)
		})
		it('should generate positive positions if startAt is after the reference origin', () => {
			expect(calculateStartingPosition(
				'2021-03-12T20:04:26.258Z',
				'2021-03-12T20:04:25.258Z',
			)).toBe(1000)
		})
	})

	describe('generateSegmentPayloadsFromPosition', () => {
		it('should generate payloads for simple segments', () => {
			const segments = [0, 1, 2, 3]
			const segmentPayloads = generateSegmentPayloadsFromPosition(0, segments)
			segmentPayloads.forEach((segmentPayload) => expect(segmentPayload).toMatchSnapshot({
				createdAt: expect.any(String),
			}))
		})
		it('should generate payloads for object segments', () => {
			const segments = [{
				offset: 0,
				data: {
					programName: 'The First Segment',
				},
			}, {
				offset: 300000,
				data: {
					programName: 'The Second Segment',
				},
			}]
			const segmentPayloads = generateSegmentPayloadsFromPosition(0, segments)
			expect(segmentPayloads.length).toBe(2)
			segmentPayloads.forEach((segmentPayload) => expect(segmentPayload).toMatchSnapshot({
				createdAt: expect.any(String),
			}))
		})
		it('should generate payloads for mixed segments', () => {
			const segments = [{
				offset: 0,
				data: {
					programName: 'The First Segment',
				},
			}, 500]
			const segmentPayloads = generateSegmentPayloadsFromPosition(0, segments)
			expect(segmentPayloads.length).toBe(2)
			segmentPayloads.forEach((segmentPayload) => expect(segmentPayload).toMatchSnapshot({
				createdAt: expect.any(String),
			}))
		})
	})

	describe('getPeriodPosition', () => {
		it('should return the starting position of a given period', () => {
			expect(getPeriodPosition(1000, 0, 3000)).toBe(0)
			expect(getPeriodPosition(4000, 0, 3000)).toBe(3000)
		})
		it('should return the origin position if the interval is infinite', () => {
			expect(getPeriodPosition(1000, 0, INTERVALS.INFINITE)).toBe(0)
			expect(getPeriodPosition(1000, 500, INTERVALS.INFINITE)).toBe(500)
		})
		it('should return the starting position of a given period with a non-zero origin', () => {
			expect(getPeriodPosition(1000, 60, 3000)).toBe(60)
			expect(getPeriodPosition(4000, 60, 3000)).toBe(3060)
			expect(getPeriodPosition(1000, -1000, 3000)).toBe(-1000)
			expect(getPeriodPosition(4000, -1000, 3000)).toBe(2000)
		})
	})

	describe('generateSegmentPayloadsForPositionRange', () => {
		it('should generate multiple sets of segments for positions that span multiple intervals', () => {
			const rangeStartPosition = 0
			const rangeEndPosition = 6000
			const startingPosition = 0
			const interval = 3000
			const segments = [0, 1000, 2000]
			const segmentPayloads = generateSegmentPayloadsForPositionRange(
				rangeStartPosition,
				rangeEndPosition,
				startingPosition,
				interval,
				segments,
			)
			expect(segmentPayloads.length).toBe(7)
			segmentPayloads.forEach((segmentPayload) => expect(segmentPayload).toMatchSnapshot({
				createdAt: expect.any(String),
			}))
		})

		it('should generate a subset of segments for positions overlap multiple intervals', () => {
			const rangeStartPosition = 2000
			const rangeEndPosition = 4000
			const startingPosition = 0
			const interval = 3000
			const segments = [0, 1000, 2000]
			const segmentPayloads = generateSegmentPayloadsForPositionRange(
				rangeStartPosition,
				rangeEndPosition,
				startingPosition,
				interval,
				segments,
			)
			expect(segmentPayloads.length).toBe(3)
			segmentPayloads.forEach((segmentPayload) => expect(segmentPayload).toMatchSnapshot({
				createdAt: expect.any(String),
			}))
		})

		it('should generate all segments if there is an infinite interval', () => {
			const rangeStartPosition = 2000
			const rangeEndPosition = 4000
			const startingPosition = 0
			const interval = INTERVALS.INFINITE
			const segments = [0, 1000, 2000]
			const segmentPayloads = generateSegmentPayloadsForPositionRange(
				rangeStartPosition,
				rangeEndPosition,
				startingPosition,
				interval,
				segments,
			)
			expect(segmentPayloads.length).toBe(1)
			segmentPayloads.forEach((segmentPayload) => expect(segmentPayload).toMatchSnapshot({
				createdAt: expect.any(String),
			}))
		})
	})
})
