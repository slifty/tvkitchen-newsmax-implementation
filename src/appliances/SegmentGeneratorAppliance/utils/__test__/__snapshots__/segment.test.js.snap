// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`segment utils #unit generateSegmentPayloadsForPositionRange should generate a subset of segments for positions overlap multiple intervals 1`] = `
Object {
  "createdAt": Any<String>,
  "data": null,
  "duration": 0,
  "origin": "",
  "position": 2000,
  "type": "SEGMENT.START",
}
`;

exports[`segment utils #unit generateSegmentPayloadsForPositionRange should generate a subset of segments for positions overlap multiple intervals 2`] = `
Object {
  "createdAt": Any<String>,
  "data": null,
  "duration": 0,
  "origin": "",
  "position": 3000,
  "type": "SEGMENT.START",
}
`;

exports[`segment utils #unit generateSegmentPayloadsForPositionRange should generate a subset of segments for positions overlap multiple intervals 3`] = `
Object {
  "createdAt": Any<String>,
  "data": null,
  "duration": 0,
  "origin": "",
  "position": 4000,
  "type": "SEGMENT.START",
}
`;

exports[`segment utils #unit generateSegmentPayloadsForPositionRange should generate all segments if there is an infinite interval 1`] = `
Object {
  "createdAt": Any<String>,
  "data": null,
  "duration": 0,
  "origin": "",
  "position": 2000,
  "type": "SEGMENT.START",
}
`;

exports[`segment utils #unit generateSegmentPayloadsForPositionRange should generate multiple sets of segments for positions that span multiple intervals 1`] = `
Object {
  "createdAt": Any<String>,
  "data": null,
  "duration": 0,
  "origin": "",
  "position": 0,
  "type": "SEGMENT.START",
}
`;

exports[`segment utils #unit generateSegmentPayloadsForPositionRange should generate multiple sets of segments for positions that span multiple intervals 2`] = `
Object {
  "createdAt": Any<String>,
  "data": null,
  "duration": 0,
  "origin": "",
  "position": 1000,
  "type": "SEGMENT.START",
}
`;

exports[`segment utils #unit generateSegmentPayloadsForPositionRange should generate multiple sets of segments for positions that span multiple intervals 3`] = `
Object {
  "createdAt": Any<String>,
  "data": null,
  "duration": 0,
  "origin": "",
  "position": 2000,
  "type": "SEGMENT.START",
}
`;

exports[`segment utils #unit generateSegmentPayloadsForPositionRange should generate multiple sets of segments for positions that span multiple intervals 4`] = `
Object {
  "createdAt": Any<String>,
  "data": null,
  "duration": 0,
  "origin": "",
  "position": 3000,
  "type": "SEGMENT.START",
}
`;

exports[`segment utils #unit generateSegmentPayloadsForPositionRange should generate multiple sets of segments for positions that span multiple intervals 5`] = `
Object {
  "createdAt": Any<String>,
  "data": null,
  "duration": 0,
  "origin": "",
  "position": 4000,
  "type": "SEGMENT.START",
}
`;

exports[`segment utils #unit generateSegmentPayloadsForPositionRange should generate multiple sets of segments for positions that span multiple intervals 6`] = `
Object {
  "createdAt": Any<String>,
  "data": null,
  "duration": 0,
  "origin": "",
  "position": 5000,
  "type": "SEGMENT.START",
}
`;

exports[`segment utils #unit generateSegmentPayloadsForPositionRange should generate multiple sets of segments for positions that span multiple intervals 7`] = `
Object {
  "createdAt": Any<String>,
  "data": null,
  "duration": 0,
  "origin": "",
  "position": 6000,
  "type": "SEGMENT.START",
}
`;

exports[`segment utils #unit generateSegmentPayloadsFromPosition should generate payloads for mixed segments 1`] = `
Object {
  "createdAt": Any<String>,
  "data": Object {
    "data": Array [
      123,
      34,
      112,
      114,
      111,
      103,
      114,
      97,
      109,
      78,
      97,
      109,
      101,
      34,
      58,
      34,
      84,
      104,
      101,
      32,
      70,
      105,
      114,
      115,
      116,
      32,
      83,
      101,
      103,
      109,
      101,
      110,
      116,
      34,
      125,
    ],
    "type": "Buffer",
  },
  "duration": 0,
  "origin": "",
  "position": 0,
  "type": "SEGMENT.START",
}
`;

exports[`segment utils #unit generateSegmentPayloadsFromPosition should generate payloads for mixed segments 2`] = `
Object {
  "createdAt": Any<String>,
  "data": null,
  "duration": 0,
  "origin": "",
  "position": 500,
  "type": "SEGMENT.START",
}
`;

exports[`segment utils #unit generateSegmentPayloadsFromPosition should generate payloads for object segments 1`] = `
Object {
  "createdAt": Any<String>,
  "data": Object {
    "data": Array [
      123,
      34,
      112,
      114,
      111,
      103,
      114,
      97,
      109,
      78,
      97,
      109,
      101,
      34,
      58,
      34,
      84,
      104,
      101,
      32,
      70,
      105,
      114,
      115,
      116,
      32,
      83,
      101,
      103,
      109,
      101,
      110,
      116,
      34,
      125,
    ],
    "type": "Buffer",
  },
  "duration": 0,
  "origin": "",
  "position": 0,
  "type": "SEGMENT.START",
}
`;

exports[`segment utils #unit generateSegmentPayloadsFromPosition should generate payloads for object segments 2`] = `
Object {
  "createdAt": Any<String>,
  "data": Object {
    "data": Array [
      123,
      34,
      112,
      114,
      111,
      103,
      114,
      97,
      109,
      78,
      97,
      109,
      101,
      34,
      58,
      34,
      84,
      104,
      101,
      32,
      83,
      101,
      99,
      111,
      110,
      100,
      32,
      83,
      101,
      103,
      109,
      101,
      110,
      116,
      34,
      125,
    ],
    "type": "Buffer",
  },
  "duration": 0,
  "origin": "",
  "position": 300000,
  "type": "SEGMENT.START",
}
`;

exports[`segment utils #unit generateSegmentPayloadsFromPosition should generate payloads for simple segments 1`] = `
Object {
  "createdAt": Any<String>,
  "data": null,
  "duration": 0,
  "origin": "",
  "position": 0,
  "type": "SEGMENT.START",
}
`;

exports[`segment utils #unit generateSegmentPayloadsFromPosition should generate payloads for simple segments 2`] = `
Object {
  "createdAt": Any<String>,
  "data": null,
  "duration": 0,
  "origin": "",
  "position": 1,
  "type": "SEGMENT.START",
}
`;

exports[`segment utils #unit generateSegmentPayloadsFromPosition should generate payloads for simple segments 3`] = `
Object {
  "createdAt": Any<String>,
  "data": null,
  "duration": 0,
  "origin": "",
  "position": 2,
  "type": "SEGMENT.START",
}
`;

exports[`segment utils #unit generateSegmentPayloadsFromPosition should generate payloads for simple segments 4`] = `
Object {
  "createdAt": Any<String>,
  "data": null,
  "duration": 0,
  "origin": "",
  "position": 3,
  "type": "SEGMENT.START",
}
`;
